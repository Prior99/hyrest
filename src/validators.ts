export interface Validation {
    error?: string;
}

export type Validator<T> = (input: T) => Validation | Promise<Validation>;

/**
 * Enforces `input` to be one of the values specified in `options`.
 *
 * @param options The options of which `input` needs to be.
 *
 * @return A Converter which checks if the input is one of `options` and returns an error if that was
 *         not the case and the `input` itself otherwise.
 */
export function oneOf<T>(...options: T[]): Validator<T> {
    return (value: T) => {
        if (typeof value === "undefined") { return {}; }
        if (!options.includes(value)) { return { error: `Not one of (${options.join(", ")}).` }; }
        return {};
    };
}

/**
 * Makes sure the given input is not `null` or `undefined`.
 *
 * @param input The input to check.
 *
 * @return The input if it was not `undefined` or `null`, otherwise an error will be returned.
 */
export function required<T>(value: T): Validation {
    if (typeof value === "undefined" || value === null) { return { error: "Missing required field." }; }
    return {};
}
