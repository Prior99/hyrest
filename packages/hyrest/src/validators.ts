import { Scope } from "./scope";

export interface Validation {
    error?: string;
}

export type Validator<T> = (input: T, scope?: Scope) => Validation | Promise<Validation>;

/**
 * Enforces `input` to be one of the values specified in `options`.
 *
 * @param options The options of which `input` needs to be.
 *
 * @return A validator which checks if the input is one of `options` and returns an error if that was
 *         not the case.
 */
export function oneOf<T>(...options: T[]): Validator<T> {
    return (value: T) => {
        if (typeof value === "undefined") { return {}; }
        if (!options.includes(value)) { return { error: `Not one of (${options.join(", ")}).` }; }
        return {};
    };
}

/**
 * Makes sure the given input is not `null` or `undefined`.
 *
 * @param input The input to check.
 *
 * @return The input if it was not `undefined` or `null`, otherwise an error will be returned.
 */
export function required<T>(value: T): Validation {
    if (typeof value === "undefined" || value === null) { return { error: "Missing required field." }; }
    return {};
}

export function length({ min, max }: { max: number, min: number }): Validator<string | any[]>;
export function length(min: number, max: number): Validator<string | any[]>;
/**
 * Validates the input string to be at least of length `min` and not longer than `max`.
 *
 * @param options An object with the keys `min` and `max` for the minimum and maximum length of the string.
 *
 * @return A validator which checks the length of the input string.
 */
export function length(arg1: { max: number, min: number } | number, arg2?: number): Validator<string | any[]> {
    const { min, max } = typeof arg1 === "object" ? arg1 : { min: arg1, max: arg2 };
    return (value: string | any[]) => {
        if (typeof value === "undefined") { return {}; }
        if (value === null) { return { error: "Cannot determine length of null." }; }
        if (typeof max !== "undefined" && value.length > max) {
            return { error: `Exceeds maximum length of ${max}.` };
        }
        if (typeof min !== "undefined" && value.length < min) {
            return { error: `Shorter than minimum length of ${min}.` };
        }
        return {};
    };
}

// Taken from http://emailregex.com/
const emailValidationRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/; // tslint:disable-line
// Taken from https://github.com/mixer/uuid-validate
const uuidValidationRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-4][0-9a-f]{3}-[0-9a-f]{4}-[0-9a-f]{12}$/;

/**
 * Validates that the input is an email.
 *
 * @param value The value to check.
 *
 * @return An error if `value` was not an email.
 */
export function email(value: string): Validation {
    if (typeof value === "undefined") { return {}; }
    if (typeof value !== "string" || !value.match(emailValidationRegex)) { // tslint:disable-line
        return { error: `String is not a valid email.` };
    }
    return {};
}

/**
 * Limits a validator to a given scope. The validator will only be executed if the scope matches.
 *
 * @param onlyScope The scope to limit the validator to.
 * @param validator The validator to execute if the scope matches.
 *
 * @return A validator only executed if the scope matches.
 */
export function only<T>(onlyScope: Scope, validator: Validator<T>): Validator<T> {
    return (value: T, currentScope: Scope) => {
        if (currentScope === onlyScope) {
            return validator(value);
        }
        return {};
    };
}

export function range({ min, max }: { max: number, min: number }): Validator<number>;
export function range(min: number, max: number): Validator<number>;
/**
 * Validates the input to be a number in the specified range.
 *
 * @param options An object with the keys `min` and `max` for the minimum and maximum length of the string.
 *
 * @return A validator which checks the range of the input string.
 */
export function range(arg1: { max: number, min: number } | number, arg2?: number): Validator<number> {
    const { min, max } = typeof arg1 === "object" ? arg1 : { min: arg1, max: arg2 };
    return (value: number) => {
        if (typeof value === "undefined") { return {}; }
        if (value === null) { return { error: "Cannot compare with null." }; }
        if (typeof max !== "undefined" && value > max) {
            return { error: `Exceeds maximum of ${max}.` };
        }
        if (typeof min !== "undefined" && value < min) {
            return { error: `Less than minimum of ${min}.` };
        }
        return {};
    };
}

/**
 * Validates that the input is an uuid.
 *
 * @param value The value to check.
 *
 * @return An error if `value` was not an uuid.
 */
export function uuid(value: string): Validation {
    if (typeof value === "undefined") { return {}; }
    if (typeof value !== "string" || !value.match(uuidValidationRegex)) {
        return { error: `String is not a valid uuid.` };
    }
    return {};
}
