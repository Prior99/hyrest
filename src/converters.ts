import { FullValidator, Processed } from "./validation";

export interface Converted<T> {
    error?: string;
    value?: T;
}

export type Converter<T> = (input: any) => Converted<T> | Promise<Converted<T>> | Processed<T> | Promise<Processed<T>>;

/**
 * Converts the given input to an integer if possible.
 *
 * @param input The input to convert to an integer. Can be anything in any form.
 *
 * @return An integer if the conversion succeeded or an error if the input was not a valid
 *         integer.
 */
export function int(input: any): Converted<number> {
    if (typeof input === "undefined") { return { value: input }; }
    const value = parseInt(input);
    if (isNaN(value) || parseFloat(input) !== value) { return { error: "Not a valid integer." }; }
    return { value };
}

/**
 * Converts the given input to a float if possible.
 *
 * @param input The input to convert to a float. Can be anything in any form.
 *
 * @return A float if the conversion succeeded or an error if the input was not a valid
 *         float.
 */
export function float(input: any): Converted<number> {
    if (typeof input === "undefined") { return { value: input }; }
    const value = parseFloat(input);
    if (isNaN(value)) { return { error: "Not a valid float." }; }
    return { value };
}

/**
 * Makes sure the given input is a string.
 *
 * @param input The input to check.
 *
 * @return The string in `input` if it was a string and an error otherwise.
 */
export function str(value: any): Converted<string> {
    if (typeof value === "undefined") { return { value }; }
    if (typeof value !== "string") { return { error: "Not a valid string." }; }
    return { value };
}

/**
 * Makes sure the given input is an object.
 *
 * @param input The input to check.
 *
 * @return The input if it was an object and an error otherwise.
 */
export function obj(value: any): Converted<Object> {
    if (typeof value === "undefined") { return { value }; }
    if (typeof value !== "object" || Array.isArray(value)) { return { error: "Not a valid object."}; }
    return { value };
}

/**
 * Makes sure the given input is a boolean.
 *
 * @param input The input to check.
 *
 * @return The boolean in `input` if it was a boolean and an error otherwise.
 */
export function bool(value: any): Converted<boolean> {
    if (typeof value === "undefined") { return { value }; }
    if (value === "true" || value === "false") { return { value: value === "true" }; }
    if (typeof value !== "boolean") { return { error: "Not a valid boolean." }; }
    return { value };
}
