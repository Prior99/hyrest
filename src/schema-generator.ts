import "reflect-metadata";
import { Schema, getValidatedProperties, getPropertyValidation, is } from "./validation";
import { Constructable } from "./types";

export function schemaFrom(from: Function): Schema {
    const properties = getValidatedProperties(from.prototype);
    return properties.reduce((result, { property, propertyType }) => {
        const options = getPropertyValidation(from.prototype, property);
        (result as any)[property] = is(options.converter)
            .validate(...options.validators)
            .validateCtx(options.validatorFactory);
        return result;
    }, {});
}
