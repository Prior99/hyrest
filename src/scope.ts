import "reflect-metadata";

import { Constructable } from "./types";
import * as invariant from "invariant";
import { allKeys } from "./all-keys";
import { getTransforms } from "./transform";
import { isCustomClass } from "./validation";

export type TypeCreator = () => Constructable<any>;

export interface SpecifiedTypes {
    /**
     * Types specified for method's parameters.
     */
    params: Map<number, TypeCreator>;
    /**
     * The type specified for the property itself.
     */
    property?: TypeCreator;
}

export interface PropertyMeta {
    /**
     * The prototype of the class on which the property was decorated.
     */
    readonly target: Object;
    /**
     * The name of the property that was decorated.
     */
    readonly property: string;
    /**
     * The expected type for this property.
     */
    readonly expectedType: any | any[];
}

export class Scope {
    private included: Scope[] = [];
    private ownProperties: PropertyMeta[] = [];

    /**
     * Include all properties from another scope in this scope:
     *
     * **Example:**
     *
     * ```
     * const exampleA = createScope();
     * const exampleB = createScope().include(exampleA);
     *
     * class A {
     *     @scope(exampleA)
     *     public property1: string;
     *
     *     @scope(exampleB)
     *     public property2: string;
     * }
     *
     * console.log(exampleA.properties); // Only `property1`.
     * console.log(exampleB.properties); // Will be `property1` and `property2`.
     * ```
     * @param scopes A list of scopes to include in this scope.
     *
     * @return This instance to use this method for the builder pattern.
     */
    public include(...scopes: Scope[]): Scope {
        scopes.forEach(includedScope => this.included.push(includedScope));
        // Builder pattern: Return `this`.
        return this;
    }

    public registerProperty(property: PropertyMeta) {
        this.ownProperties.push(property);
    }

    /**
     * All properties decorated by `@scope` with this scope and all properties
     * included by using `include()`.
     */
    public get properties() {
        return this.included.reduce((result, included) => {
            result.push(...included.properties);
            return result;
        }, [ ...this.ownProperties ]);
    }

    /**
     * Will return all properties for a specific class.
     *
     * **Example:**
     *
     * ```
     * const example = createScope();
     *
     * class A {
     *     @scope(example)
     *     public property1: string;
     * }
     *
     * class B {
     *     @scope(example)
     *     public property2: string;
     * }
     *
     * console.log(example.propertiesForClass(A)); // Only `property1` is returned.
     * console.log(example.propertiesForClass(A)); // Only `property2` is returned.
     * ```
     *
     * This method takes all properties included using `include()` into account.
     *
     * @param clazz The clazz to return the properties for.
     *
     * @return All properties from the specified class.
     */
    public propertiesForClass(clazz: Function) {
        return this.properties.filter(property => {
            let current = clazz;
            while (current !== null && typeof current !== "undefined") { // tslint:disable-line
                if (property.target.constructor === current) {
                    return true;
                }
                current = Object.getPrototypeOf(current);
            }
        });
    }
}

/**
 * Create a new scope.
 *
 * @return A new scope.
 */
export function createScope() {
    return new Scope();
}

/**
 * Use this decorator to mark a property, getter or a setter on a class as belonging to a scope
 * so they are included when `dump()`ing or `populate()`ing them.
 *
 * @param scopes The list of scopes in which the decorated property should be included.
 *
 * @return A decorator which can be used to decorate getters, setters and properties.
 *
 * @see dump
 * @see populate
 */
export function scope(...scopes: Scope[]) {
    return function<T>(target: Object, property: string, descriptor?: PropertyDescriptor) {
        const expectedType = Reflect.getMetadata("design:type", target, property);
        scopes.forEach(decoratedScope => decoratedScope.registerProperty({
            target, property, expectedType,
        }));
    };
}

/**
 * Returns the specified types for a property and its parameters. Is always guaranteed to
 * return an object. If non existed, a new one is created.
 *
 * @param target The target object of which to get the specified types.
 * @param property The property on `target` on which to get the specified types.
 *
 * @return An object containing the type of the property and a map for all its parameters.
 */
export function getSpecifiedType(target: Object, property: string | symbol): SpecifiedTypes {
    let specified = Reflect.getMetadata("specifytype", target, property);
    if (typeof specified === "undefined") {
        specified = {
            params: new Map(),
        };
        Reflect.defineMetadata("specifytype", specified, target, property);
    }
    return specified;
}

/**
 * Help to specify the type of a property or paramter. Might be necessary for arrays and cyclic
 * depdendencies.
 *
 * @param factory A function with no arguments returning the desired type.
 *
 * @return A decorator for a property or a parameter.
 */
export function specify<T>(factory: TypeCreator) {
    return function(target: Object, property: string | symbol, arg3?: any) {
        const specified = getSpecifiedType(target, property);
        if (typeof arg3 === "number") {
            specified.params.set(arg3, factory);
            return;
        }
        specified.property = factory;
    };
}

export function dump<T>(dumpScope: Scope): (instance: T) => T;
export function dump<T>(dumpScope: Scope, instance: T): T;
/**
 * Dump a structure with a given scope. This can either be invoked using the curried syntax
 * (`dump(scope)(obj)` or regularly (`dump(scope, obj)`).
 * This will dump the structure as new, plain objects, including only the keys marked with the
 * `@scope` decorator and the specified scope.
 *
 * @param dumpScope The scope used for dumping.
 * @param arg2 The instance to dump.
 *
 * @return The dumped, stripped down version of the input object.
 */
export function dump<T>(dumpScope: Scope, arg2?: T): T | ((instance: T) => T) {
    function internalDump<U extends any | any[]>(instance: U): U {
        if (Array.isArray(instance)) {
            return instance.map(internalDump) as any as U;
        }
        if (typeof instance === "object" && instance !== null) {
            const propertiesForClass = dumpScope.propertiesForClass(instance.constructor);
            const keys = allKeys(instance);
            return keys.reduce((result, key) => {
                if (propertiesForClass.find(({ property }) => property === key)) {
                    (result as any)[key] = internalDump((instance as any)[key]);
                }
                return result;
            }, {}) as U;
        }
        return instance;
    }
    if (typeof arg2 !== "undefined") {
        return internalDump(arg2);
    }
    return internalDump;
}

export function populate<T>(populateScope: Scope, initialClass: Constructable<T>): (data: any) => T;
export function populate<T>(populateScope: Scope, initialClass: Constructable<T>, data: any): T;
export function populate<T>(
    populateScope: Scope, initialClass: Constructable<T>, arrayType: Constructable<T>): (data: any) => T;
export function populate<T>(
    populateScope: Scope, initialClass: Constructable<T>, arrayType: Constructable<T>, data: any): T;
/**
 * Populate a structure of classes using an object and a scope.
 * This will recursively create instances of all classes and populate them with the nested objects in
 * the input.
 *
 * @param populateScope The scope to populate. Only properties included in this scope will be populated.
 * @param initialClass The class to start populating with.
 *             It can also be invoked with an additional array type.
 * @param arg4 When populating an array, an additional array type must be specified as the thrid argument
 *             and the data is specified as the fourth argument instead.
 *
 * @return A curried function for use in higher order functions if `arg3` is not specified and the populated
 *         structure otherwise.
 */
export function populate<T>(
    populateScope: Scope, initialClass: Constructable<T>, arg3?: any, arg4?: any,
): T | ((data: T) => T) {
    function internalPopulate<U extends any | any[]>(
        data: any, thisClass: any, specifiedClass: Constructable<any>,
    ): U {
        // Perform population of an array.
        if (thisClass === Array) {
            invariant(Array.isArray(data), "Structure does not match. Array expected.");
            invariant(typeof specifiedClass === "function", "Array type not specified.");
            return (data as any[]).map(element => internalPopulate(element, specifiedClass, undefined)) as any as U;
        }
        const guardedClass = thisClass || specifiedClass;
        // Ignore primitives.
        if (guardedClass === Number || guardedClass === Boolean || guardedClass === String || guardedClass === Object) {
            return data;
        }
        invariant(typeof data === "object" && !Array.isArray(data), "Structure does not match. Object expected.");
        if (typeof guardedClass !== "function") {
            throw new Error("Could not infer type. This might be due to a cyclic dependency.");
        }
        // Instanciate the classe to populate.
        const instance = new guardedClass();
        // Get a list of all properties to populate.
        const propertiesForClass = populateScope.propertiesForClass(guardedClass);
        propertiesForClass.forEach(({ property, target, expectedType }) => {
            const dataValue = (data as any)[property];
            if (typeof dataValue === "undefined") {
                return;
            }
            const specifyTypeCreator = getSpecifiedType(target, property).property;
            const specifyType = specifyTypeCreator && specifyTypeCreator();
            const transforms = getTransforms(target, property);
            const populated = internalPopulate(dataValue, expectedType, specifyType);
            if (transforms.propertyTransform) {
                (instance as any)[property] = transforms.propertyTransform(populated);
            } else {
                (instance as any)[property] = populated;
            }
        });
        return instance;
    }
    // `arg3` can either be array type or data.
    if (typeof arg3 !== "undefined") {
        // `arg3 is the array type and `arg4` is the data.
        if (typeof arg4 !== "undefined") {
            return internalPopulate(arg4, initialClass, arg3);
        }
        // If `arg3` is a `Constructable`: It's the array type and a curried version shall be returned.
        if (typeof arg3 === "function") {
            return (data: any) => internalPopulate(data, initialClass, arg3);
        }
        // `arg3 is simply the data and no array type is specified.
        return internalPopulate(arg3, initialClass, undefined);
    }
    return (data: any) => internalPopulate(data, initialClass, undefined);
}
