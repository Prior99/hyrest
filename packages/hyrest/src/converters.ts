import { FullValidator } from "./validation";
import { Processed } from "./processed";
import { Scope } from "./scope";

export interface Converted<T> {
    error?: string;
    value?: T;
}

export type Converter<T> = (input: any, scope?: Scope, context?: any) =>
    Converted<T> |
    Promise<Converted<T>> |
    Processed<T> |
    Promise<Processed<T>>;

/**
 * Converts the given input to an integer if possible.
 *
 * @param input The input to convert to an integer. Can be anything in any form.
 *
 * @return An integer if the conversion succeeded or an error if the input was not a valid
 *         integer.
 */
export function int(input: any): Converted<number> {
    if (typeof input === "undefined") { return { value: input }; }
    const value = parseInt(input);
    if (isNaN(value) || parseFloat(input) !== value) { return { error: "Not a valid integer." }; }
    return { value };
}

/**
 * Converts the given input to a float if possible.
 *
 * @param input The input to convert to a float. Can be anything in any form.
 *
 * @return A float if the conversion succeeded or an error if the input was not a valid
 *         float.
 */
export function float(input: any): Converted<number> {
    if (typeof input === "undefined") { return { value: input }; }
    const value = parseFloat(input);
    if (isNaN(value)) { return { error: "Not a valid float." }; }
    return { value };
}

/**
 * Makes sure the given input is a string.
 *
 * @param input The input to check.
 *
 * @return The string in `input` if it was a string and an error otherwise.
 */
export function str(value: any): Converted<string> {
    if (typeof value === "undefined") { return { value }; }
    if (typeof value !== "string") { return { error: "Not a valid string." }; }
    return { value };
}

/**
 * Makes sure the given input is an object.
 *
 * @param input The input to check.
 *
 * @return The input if it was an object and an error otherwise.
 */
export function obj(value: any): Converted<Object> {
    if (typeof value === "undefined") { return { value }; }
    if (typeof value !== "object" || Array.isArray(value)) { return { error: "Not a valid object."}; }
    return { value };
}

/**
 * Makes sure the given input is a boolean.
 *
 * @param input The input to check.
 *
 * @return The boolean in `input` if it was a boolean and an error otherwise.
 */
export function bool(value: any): Converted<boolean> {
    if (typeof value === "undefined") { return { value }; }
    if (value === "true" || value === "false") { return { value: value === "true" }; }
    if (typeof value !== "boolean") { return { error: "Not a valid boolean." }; }
    return { value };
}

/**
 * Makes sure the given input is an array with all elements matching the given criteria.
 *
 * @param validator A validator to apply to all element of the array.
 *
 * @return The array converter.
 */
export function arr<T, TContext>(validator?: FullValidator<T, TContext>): Converter<T[]> {
    return async (value: any, scope?: Scope, context?: any) => {
        const processed = new Processed<T[]>();
        // Ignore `undefined` inputs. This is handled by the `required` validator if intended by
        // the user.
        if (typeof value === "undefined") { return processed; }
        // If the input is not an array at all, don't even try to validate any elements.
        if (!Array.isArray(value)) {
            processed.addErrors("Not an array.");
            return processed;
        }
        // Apply the validator (if provided) to all elements and store the results in the `nested`
        // property of the result. The keys will be the array indices.
        if (typeof validator !== "undefined"){
            const results = await Promise.all(value.map(async (elem, index) => {
                const result = await validator(elem, { scope, context });
                if (result.hasErrors) {
                    processed.addNested(index, result);
                }
                return result;
            }));
            if (!results.find(result => result.hasErrors)) {
                processed.value = results.map(result => result.value);
            }
            return processed;
        }
        processed.value = value;
        return processed;
    };
}

/**
 * Makes sure the given input is `null`.
 *
 * @param input The input to check.
 *
 * @return `null` if it was a `null` and an error otherwise.
 */
export function empty(value: any): Converted<null> {
    if (value === null) { return { value }; }
    return { error: "Not null." };
}
