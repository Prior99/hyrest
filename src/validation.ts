import "reflect-metadata";

import { Validator, Validation } from "./validators";
import { schemaFrom } from "./schema-generator";
import { Converter, arr, bool, str, float, obj } from "./converters";

export interface ValidationOptions<T> {
    converter?: Converter<T>;
    readonly validators: Validator<T>[];
    validatorFactory?: (ctx: any) => Validator<T>[];
}

export interface Schema {
    [key: string]: Schema | FullValidator<any>;
}

export interface Processed<T> {
    value?: T;
    errors: string[];
}

async function validateSchema<T extends Object>(validationSchema: Schema, input: T): Promise<boolean> {
    const result = await Promise.all(Object.keys(validationSchema).map(async key => {
        const schemaValue = validationSchema[key];
        const inputValue = (input as any)[key];

        if (typeof schemaValue === "function") {
            return (await schemaValue(inputValue)).errors.length === 0;
        }

        return await validateSchema(schemaValue, inputValue);
    }));
    const noAdditionalKeys = Object.keys(input).every(key => Object.keys(validationSchema).includes(key));
    return result.every(keyResult => keyResult) && noAdditionalKeys;
}

export function schema<T extends Object>(validationSchema: Schema): Validator<T> {
    return async (value: T) => {
        if (typeof value === "undefined") {
            return {};
        }
        return (await validateSchema(validationSchema, value)) ? {} : { error: "Schema validation failed." };
    };
}

/**
 * Fully processes an unprocessed input value. This means conversion and validation of the value.
 *
 * @param input The input to process.
 * @param converter A function which will convert the input value into a specific type.
 * @param validators A list of validators to validate the input value.
 *
 * @return An object containing all errors and the converted value.
 */
export async function processValue<T>(
        input: any, converter: Converter<T>, validators: Validator<T>[]): Promise<Processed<T>> {
    const { error, value } = converter ?
            await converter(input) :
            { value: input, error: undefined };
    if (error) {
        return {
            errors: [error],
        };
    }
    const errors = (await Promise.all(validators.map(validator => validator(value))))
        .reduce((result, { error: validationError }) => {
            if (validationError) {
                result.push(validationError);
            }
            return result;
        }, []);
    return { value, errors };
}

type ValidationMap = Map<number, ValidationOptions<any>>;

/**
 * Retrieves all the validators and the converter for a given parameter at index `index` on the method with name
 * `propertyKey` on the instance of a controller `target`.
 * This function is always guaranteed to return the options. If nothing was defined yet, a new metadata key with
 * an empty options object will be defined.
 *
 * @param target The target instance of an object on which a method's parameter should be looked up.
 * @param propertyKey The name of a method on `target` on which a parameter should be decorated.
 * @param index The index of the parameter to decorate. This argument is optional. If spared, the validation is
 *              appended to the property instead of the parameter.
 *
 * @return An options object to which new validators and a converter can be appended. Is always guaranteed to return
 *         an options object.
 */
export function getParameterValidation(
    target: Object, propertyKey: string | symbol, index: number,
): ValidationOptions<any> {
    // Try to retrieve the `Map` of options with the keys being the parameter index and the value being
    // an the options object.
    const map: ValidationMap = Reflect.getMetadata("validation:parameters", target, propertyKey);

    // If no map has been found then this function has never been called for this method before. A new map needs
    // to be created and an empty options object needs to be attached.
    if (!map) {
        const newMap = new Map<number, ValidationOptions<any>>();
        const newOptions: ValidationOptions<any> = { validators: [] };
        newMap.set(index, newOptions);

        // Define the new key on the reflection metadatas.
        Reflect.defineMetadata("validation:parameters", newMap, target, propertyKey);
        return newOptions;
    }

    // If the code reaches here, a map existed and at least something has already been attached to the map.
    const options = map.get(index);

    // If no options are present, a new object needs to be created and inserted into the map.
    if (!options) {
        const newOptions: ValidationOptions<any> = { validators: [] };
        map.set(index, newOptions);
        return newOptions;
    }
    return options;
}

/**
 * Returns the validation options defined on the specified property. This will always return an
 * array. If no validations options had been defined previously, a new array is created into
 * which new options can be added.
 *
 * @param target The target instance of an object on which the property is defined.
 * @param propertyKey The name of a property on `target` which hould be decorated.
 *
 * @return An array of options objects to which new validators and a converter can be appended.
 * Is always guaranteed to return an array.
 */
export function getPropertyValidation(target: Object, propertyKey: string): ValidationOptions<any> {
    const options = Reflect.getMetadata("validation:property", target, propertyKey);
    if (!options) {
        const newOptions: ValidationOptions<any> = { validators: [] };
        Reflect.defineMetadata("validation:property", newOptions, target, propertyKey);
        return newOptions;
    }
    return options;
}

export interface ValidatedProperty {
    readonly property: string;
    readonly propertyType: Function;
}

export function getValidatedProperties(target: Object): ValidatedProperty[] {
    const properties = Reflect.getMetadata("validation:properties", target);
    if (!properties) {
        const newProperties: ValidatedProperty[] = [];
        Reflect.defineMetadata("validation:properties", newProperties, target);
        return newProperties;
    }
    return properties;
}

export interface FullValidator<T> {
    (input: any): Processed<T> | Promise<Processed<T>>;
    (target: Object, propertyKey: string | symbol, index: number): void;
    validate: (...validators: Validator<T>[]) => FullValidator<T>;
    validateCtx: (factory: (ctx: any) => Validator<T>[]) => FullValidator<T>;
    validators: Validator<T>[];
    validatorFactory: (ctx: any) => Validator<T>[];
}

/**
 * This decorator can be applied to a parameter in a `@route` and will make sure that the input
 * can be converted to the given data type and will then convert it.
 *
 * @param converter A converter to convert the raw input.
 *
 * @return A decorator for a parameter in a @route method.
 */
export function is<T>(converter?: Converter<T>): FullValidator<T> {
    const fn: any = (...args: any[]) => {
        if (args.length !== 3) {
            // Called as a function.
            const factoryValidators = fn.validatorFactory ? fn.ValidatorFactory(this) : []; //tslint:disable-line
            return processValue(args[0], converter, [...fn.validators, ...factoryValidators]);
        }
        else if (typeof args[2] === "number") {
            // Parameter decorator.
            const options = getParameterValidation(args[0], args[1], args[2]);
            options.converter = converter;
            options.validators.push(...fn.validators);
            options.validatorFactory = fn.validationFactory;
            return;
        } else {
            const propertyType = Reflect.getMetadata("design:type", args[0], args[1]);
            const arrayOfType = Reflect.getMetadata("arrayof", args[0], args[1]);
            getValidatedProperties(args[0]).push({
                property: args[1],
                propertyType,
            });
            // Property decorator.
            const options = getPropertyValidation(args[0], args[1]);
            options.converter = typeof converter === "function" ?
                converter :
                inferConverter(propertyType, arrayOfType);
            options.validators.push(...fn.validators);
            options.validatorFactory = fn.validationFactory;
            return;
        }
    };
    fn.validators = [];
    fn.validate = (...validators: Validator<T>[]) => {
        fn.validators.push(...validators);
        return fn;
    };
    fn.validateCtx = (factory: (ctx: any) => Validator<T>[]) => {
        fn.validationFactory = factory;
        return fn;
    };
    return fn as FullValidator<T>;
}

/**
 * Infers the needed converter based on the constructor type.
 *
 * @param ctor The constructor such as `Number` or `Object`.
 *
 * @return The corresponding converter.
 */
export function inferConverter(ctor: Function, arrayOfType?: Function): Converter<any> {
    if (ctor === Number) {
        return float;
    }
    if (ctor === String) {
        return str;
    }
    if (ctor === Boolean) {
        return bool;
    }
    if (ctor === Object) {
        return obj;
    }
    if (ctor === Array) {
        if (arrayOfType) {
            return arr(is(inferConverter(arrayOfType)));
        }
        return arr();
    }
    if (ctor === Function) {
        return schema(schemaFrom(ctor));
    }
}
