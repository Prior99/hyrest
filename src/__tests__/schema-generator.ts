import { schemaFrom } from "../schema-generator";
import { createScope, scope, arrayOf, Scope } from "../scope";
import { is, validateSchema } from "../validation";
import { email, oneOf, required } from "../validators";
import { int, str, obj, arr } from "../converters";

let A, B: Function;
let scope1, scope2: Scope;

beforeEach(() => {
    scope1 = createScope();
    scope2 = createScope();

    class _A { //tslint:disable-line
        @scope(scope1)
        @is(str).validate(email, required)
        public email: string;

        @scope(scope2)
        @is(int).validate(oneOf(1, 2, 3))
        public anInteger: number;
    }

    class _B { //tslint:disable-line
        @scope(scope1) @is().validate(required)
        public a: _A;

        @scope(scope2) @is() @arrayOf(_A)
        public as: _A[];

        @scope(scope1, scope2) @is() @arrayOf(_B)
        public bs: _B[];
    }

    A = _A;
    B = _B;
});

[
    {
        a: { email: "test@example.com", anInteger: 1 },
        bs: [
            {
                a: { email: "some@example.com" },
                bs: [],
                as: [ { email: "some@example.com", anInteger: 3 }, { email: "an@example.com" } ],
            },
        ],
    },
    {
        a: { email: "foo@example.com" },
        bs: [],
    },
].forEach((input, index) => {
    test(`the generated schema detects valid inputs as valid (${index})`, async () => {
        expect(await validateSchema(schemaFrom(B), input)).toEqual({});
    });
});

[
    {
        a: { email: "test@example.com" },
        bs: [
            {
                a: { email: "some@example.com" },
                bs: [],
            },
        ],
    },
    {
        a: { email: "foo@example.com" },
        bs: [],
    },
].forEach((input, index) => {
    test(`the generated schema detects valid inputs as valid with a scope supplied (${index})`, async () => {
        expect(await is(obj).schema(schemaFrom(B)).scope(scope1)(input)).toMatchSnapshot();
    });
});

[
    {
        a: { email: "test@example.com" },
        bs: [],
        as: [],
    },
    {
        a: { email: "foo@example.com", anInteger: 1 },
    },
    {
        a: { email: "invalid-email" },
    },
].forEach((input, index) => {
    test(`the generated schema detects invalid inputs as invalid with a scope supplied (${index})`, async () => {
        expect(await is(obj).schema(schemaFrom(B)).scope(scope1)(input)).toMatchSnapshot();
    });
});

[
    {},
    {
        bs: [],
    },
    {
        a: { email: "invalid-email" },
    },
    {
        a: { email: "some@exmaple.com", anInteger: 3.141 },
    },
    {
        a: { email: "some@exmaple.com", anInteger: 2 },
        bs: {
            a: { email: "some@exmaple.com", anInteger: 3 },
            bs: [],
        },
    },
    {
        a: { email: "some@exmaple.com", anInteger: 3 },
        bs: [
            {
                a: { email: "some@exmaple.com", anInteger: 3 },
                bs: [],
            },
        ],
        as: [
            {},
        ],
    },
    {
        a: { email: "some@exmaple.com", anInteger: 2 },
        bs: [ { dangling: "invalid" } ],
    },
].forEach((input, index) => {
    test(`the generated schema detects invalid inputs as invalid (${index})`, async () => {
        const result = await validateSchema(schemaFrom(B), input);
        expect(result).toMatchSnapshot();
        expect(result.hasErrors).toBe(true);
    });
});

test("throws an error if decorated incorrectly", () => {
    expect(() => {
        class C { // tslint:disable-line
            @arrayOf(str) @is()
            public test: string[];
        }
    }).toThrowErrorMatchingSnapshot();
});
