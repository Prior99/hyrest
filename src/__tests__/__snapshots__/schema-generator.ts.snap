// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the generated schema is the same as a manually defined schema 1`] = `
Object {
  "anInteger": [Function],
  "email": [Function],
}
`;

exports[`the generated schema is the same as a manually defined schema 2`] = `
Object {
  "a": [Function],
  "as": [Function],
  "bs": [Function],
}
`;
